runtime/j9vm/jvm.c

```diff
@@ -3989,15 +3989,79 @@ JVM_LoadLibrary(const char *libName, jboolean throwOnFailure)
- */
-#if JAVA_SPEC_VERSION < 17
 			if (0 != slOpenResult) {
-				slOpenResult = j9sl_open_shared_library((char *)libName, &handle, flags | J9PORT_SLOPEN_DECORATE);
-				Trc_SC_LoadLibrary_OpenShared_Decorate(libName);
+				char *libNameNotDecorated = (char *)libName;
+#if JAVA_SPEC_VERSION >= 17
+				/* JDK17+ jdk.internal.loader.NativeLibraries.load() calls JVM_LoadLibrary() with a decorated library name,
+				 * i.e., a path to the library name returned from Java_java_lang_System_mapLibraryName(nameNoPrefixNoExtension).
+				 * The library name passed to j9sl_open_shared_library() with the flag J9PORT_SLOPEN_DECORATE
+				 * has to be platform independent, i.e., it must not contain any prefix or file extension.
+				 */
+				char libPath[EsMaxPath] = {0};
+				const char *fileExt = strrchr(libName, '.');
+				BOOLEAN doOpenLibrary = TRUE;
+				if (NULL != fileExt) {
+					const size_t libSuffixLength = strlen(fileExt);
+					const char *fileNameTmp = strrchr(libName, DIR_SEPARATOR);
+					const char *fileName = (NULL == fileNameTmp) ? libName : fileNameTmp + 1;
+#if defined(J9OS_I5) || defined(WIN32)
+					/* no library prefix for J9OS_I5 and WIN32 */
+					const size_t libStrLength = 0;
+#else /* defined(J9OS_I5) || defined(WIN32) */
+					/*  strlen("lib") = 3 */
+					const size_t libStrLength = 3;
+					if (0 != strncmp("lib", fileName, libStrLength)) {
+						/* A decorated library name is expected to start with lib prefix for platforms other than WIN32 & J9OS_I5.
+						 * pass to j9sl_open_shared_library w/o modification.
+						 */
+						Trc_SC_libName_no_prefix(fileName);
+					} else
+#endif /* defined(J9OS_I5) || defined(WIN32) */
+					{
+						const char *fileNameNoPrefix = fileName + libStrLength;
+						if (NULL != fileNameTmp) {
+							/* the libName specifies a path */
+							UDATA libDirLength = (uintptr_t)fileName - (uintptr_t)libName;
+							UDATA fileNameNotDecoratedLength = strlen(fileNameNoPrefix) - libSuffixLength;
+							UDATA libPathLength = libDirLength + fileNameNotDecoratedLength + 1;
+							if (libPathLength <= EsMaxPath) {
+								j9str_printf(PORTLIB, libPath, EsMaxPath, "%.*s%.*s", libDirLength, libName,
+										fileNameNotDecoratedLength, fileNameNoPrefix);
+								libNameNotDecorated = libPath;
+							} else {
+								libNameNotDecorated = (char *)j9mem_allocate_memory(libPathLength, OMRMEM_CATEGORY_VM);
+								if (NULL != libNameNotDecorated) {
+									j9str_printf(PORTLIB, libNameNotDecorated, libPathLength, "%.*s%.*s", libDirLength, libName,
+											fileNameNotDecoratedLength, fileNameNoPrefix);
+								} else {
+									doOpenLibrary = FALSE;
+									Trc_SC_allocate_memory_failed(libPathLength);
+								}
+							}
+						} else {
+							/* the max file name length is smaller than EsMaxPath */
+							j9str_printf(PORTLIB, libPath, EsMaxPath, "%.*s", strlen(fileNameNoPrefix) - libSuffixLength, fileNameNoPrefix);
+							libNameNotDecorated = libPath;
+						}
+					}
```

Comment:
* this need not make a special case for `libDirLength == 0`
* there's no need to use `strlen()` to compute the base name length
* this should not mix `UDATA` with `uintptr_t`
```c
						const char *fileNameNoPrefix = fileName + libStrLength;
						uintptr_t libDirLength = (uintptr_t)fileName - (uintptr_t)libName;
						uintptr_t fileNameNotDecoratedLength = (uintptr_t)fileExt - (uintptr_t)fileNameNoPrefix;
						size_t libPathLength = libDirLength + fileNameNotDecoratedLength + 1;
						if (libPathLength <= EsMaxPath) {
							libNameNotDecorated = libPath;
						} else {
							libNameNotDecorated = (char *)j9mem_allocate_memory(libPathLength, OMRMEM_CATEGORY_VM);
						}
						if (NULL == libNameNotDecorated) {
							doOpenLibrary = FALSE;
							Trc_SC_allocate_memory_failed(libPathLength);
						} else {
							j9str_printf(PORTLIB,
									libNameNotDecorated,
									libPathLength,
									"%.*s%.*s",
									libDirLength,
									libName,
									fileNameNotDecoratedLength,
									fileNameNoPrefix);
						}
```

Reply:
Modified accordingly.